src/metabase/models/query_field.clj

```diff
@@ -1,9 +1,69 @@
+
+(defn- field-ids-for-mbql
+  "Find out ids of all fields used in a query. Conforms to the same protocol as [[query-analyzer/field-ids-for-sql]],
+  so returns `{:direct #{...int ids}}` map.
+
+  Does not track wildcards for queries rendered as tables afterwards."
+  [query]
+  {:direct (mbql.u/referenced-field-ids query)})
+
+(defn update-query-fields-for-card!
+  "Clears QueryFields associated with this card and creates fresh, up-to-date-ones.
+
+  Any card is accepted, but this functionality only works for ones with a native query.
+
+  If you're invoking this from a test, be sure to turn on [[*parse-queries-in-test?*]].
+
+  Returns `nil` (and logs the error) if there was a parse error."
+  [{card-id :id, query :dataset_query}]
+  (when query
+    (try
+      (let [{:keys [direct indirect] :as res} (case (:type query)
+                                                :native (try
+                                                          (query-analyzer/field-ids-for-sql query)
+                                                          (catch Exception e
+                                                            (log/error e "Error parsing SQL" query)))
+                                                :query  (field-ids-for-mbql query)
+                                                nil     nil)
+            id->row                           (fn [direct? field-id]
+                                                {:card_id          card-id
+                                                 :field_id         field-id
+                                                 :direct_reference direct?})
+            query-field-rows                  (concat
+                                               (map (partial id->row true) direct)
+                                               (map (partial id->row false) indirect))]
+        ;; when response is `nil`, it's a disabled parser, not unknown columns
+        (when (some? res)
+          (t2/with-transaction [_conn]
+            (let [existing            (t2/select :model/QueryField :card_id card-id)
+                  {:keys [to-update
+                          to-insert
+                          to-delete]} (u/row-diff existing query-field-rows :field_id [:id :card_id :field_id])]
+              (when (seq to-delete)
+                ;; this delete seems to break transaction (implicit commit or something) on MySQL, and this `diff`
+                ;; algo drops its frequency by a lot - which should help with transactions affecting each other a
+                ;; lot. Parallel tests in `metabase.models.query.permissions-test` were breaking when delete was
+                ;; executed unconditionally on every query change.
+                (t2/delete! :model/QueryField :card_id card-id :field_id [:in (map :field_id to-delete)]))
+              (when (seq to-insert)
+                (t2/insert! :model/QueryField to-insert))
+              (doseq [item to-update]
```

Comment:
Rather than complicating `row-diff` with the `non-data-keys`, one idea could be to do the comparison between existing rows here:
`(when (not= (dissoc item :id :card_id :field_id) (dissoc to-update :id :card_id :field_id)) ....)`

Reply:
Yeah, but you need to _match_ those fields (like get the one from old rows), which adds some (almost irrelevant to what it's doing) logic on the user's side, so it feels like it's nice to encapsulate it in a function. 

Could be a separate function, of course... But it seems to me that API of this function can impact a lot how results are used afterwards, and when it explicitly asks you - what do I compare? - makes you actually pay attention to that.